src/rpp/rpp/operators/combine_latest.hpp

```diff
@@ -0,0 +1,173 @@
+        {
+            if (!subscriber.is_subscribed()) return;
+
+            std::scoped_lock<std::mutex> lock{m_mutex};
+
+            if (++m_completed_count == m_total_completed_count)
+            {
+                subscriber.on_completed();
+            }
+        }
+
+    private:
+        TCombiner m_combiner;
+
+        std::mutex m_mutex;
+        std::tuple<std::optional<SourceValueType>, std::optional<OtherValueTypes>...> m_values{};
+        size_t m_completed_count{0};
+
+        const std::size_t m_total_completed_count = 1 + sizeof...(OtherValueTypes);
+
+        bool has_values_for_observer()
+        {
+            return std::apply([](const auto&... cached_values)
+            {
+                // Variadic expansion used
+                return (cached_values.has_value() && ...);
+            }, m_values);
+        }
+
+        auto combine_cached_values()
+        {
+            return std::apply([&](const auto&... cached_values)
+            {
+                // Variadic expansion used
+                return m_combiner(cached_values.value()...);
+            }, m_values);
+        }
+    };
+
+    /**
+     * \brief Templated helper function for subscribing to variadic 'other' observables.
+     *
+     * \param observables is a variadic tuple. The first observable is the current observable and rest are the other observables.
+     * \param subscriber is the downstream subscriber.
+     * \param coordinator manages the cache of emission from the observables and coordinate dispatching.
+     */
+    template<size_t...I>
+    static void subscribe_other_observables(std::index_sequence<I...>, // Used in compile time for variadic expansion
+                                            const auto& observables,
+                                            const auto& subscriber,
+                                            const auto& coordinator)
+    {
+        // +1 because the first element in tuple is the current observable, and you want to subscribe to the 'other' observables.
+        // (Use variadic expansion to iterate the observables)
+        (subscribe_observable<I + 1>(std::get<I>(observables), subscriber, coordinator), ...);
+    }
+
+    template<size_t I, constraint::observable TObs>
+    static void subscribe_observable(const TObs& observable, const auto& subscriber, const auto& coordinator)
+    {
+        using ValueType = utils::extract_observable_type_t<TObs>;
+        observable.subscribe(create_subscriber_with_state<ValueType>(
+            subscriber.get_subscription().make_child(),
+            subscriber,
+            [coordinator](auto&& value, const auto& subscriber)
+            {
+                coordinator->template offer_next_value<I>(std::forward<decltype(value)>(value), subscriber);
+            },
+            utils::forwarding_on_error{},
+            [coordinator](const auto& subscriber)
+            {
+                coordinator->offer_completed(subscriber);
+            }));
```

Comment:
Subscriber there and inside `operator()` are same. Extract function for creation of subscriber and call it there and inside `operator()`

Reply:
Oh can you provide me more snippet? I needed this function for the variadic expansion for `template<size_t...I>`.

Reply:
```cpp
template<typename ValueType, size_t I>
auto create_inner_subscriber(auto&& subscriber, const auto& coordinator)
{
auto sub = subscriber.get_subscription().make_child();
return create_subscriber_with_state<ValueType>(
            std::move(sub),
            std::forward<decltype(subscriber)>(subscriber),
            [coordinator](auto&& value, const auto& subscriber)
            {
                coordinator->template offer_next_value<I>(std::forward<decltype(value)>(value), subscriber);
            },
            utils::forwarding_on_error{},
            [coordinator](const auto& subscriber)
            {
                coordinator->offer_completed(subscriber);
            })
}


 static void subscribe_observable(const TObs& observable, const auto& subscriber, const auto& coordinator)
    {
        using ValueType = utils::extract_observable_type_t<TObs>;
        observable.subscribe(create_inner_subscriber<ValueType, I>(subscriber, coordinator));
    }


public:

template<constraint::subscriber_of_type<DownstreamType> TSub>
    auto operator()(TSub&& subscriber) const
    {
        // TODO: Avoid copying the combiner?
        auto coordinator = std::make_shared<combine_latest_coordinator_t<Type, utils::extract_observable_type_t<TObservable>...>>(m_combiner);

        // Subscribe to other observables and redirect on_next event to coordinator
        subscribe_other_observables(std::index_sequence_for<TObservable...>{}, m_other_observables, subscriber, coordinator);

       return create_inner_subscriber<ValueType, INDEX_OF_SOURCE_TYPE>(std::forward<TSub>(subscriber), coordinator);
    }
```

Reply:
Hmm, I don't think the code above will work. Can we keep the current code?

Reply:
Oh can I keep this pattern for better readability? Because not all the IDE(s) tell you the argument's name, it's kind of readable by having a variable name. Also, I believe all the compilers will optimize it by lifting the variable as unnamed variable?

```
auto on_next = [] {};
auto on_completed = [] {};

return create_subscriber_with_state<ValueType>(
    ...,
    std::move(on_next),
    ...,
    std::move(on_completed));
```

Reply:
But why? your code contains long duplicate and it is better to extract this into separate function to show that ALL inner subscribers are same

Reply:
> Hmm, how does the variadic expansion work for the variadic observables (this one + others)?

Not sure if i get your question, there is no any variadic...

Reply:
Whole this comment about extracting `create_inner_subscriber` as separate function

Reply:
> Oh can I keep this pattern for better readability? Because not all the IDE(s) tell you the argument's name, it's kind of readable by having a variable name. Also, I believe all the compilers will optimize it by lifting the variable as unnamed variable?

Not sure if is really needed. Pattern of creation of all operators are the same. Same about creation of inner subscribers - subscribtion/state/on_next/on_error/on_completed. Moreover, it is fully internal thing and some "non-experienced" user not need to read this code, while some "experienced" can understand it without named variables